use C++ std::stoi et al instead of boost::lexical_cast

#### Summary:

To avoid compiler warnings, wse the more efficient `std::stoi`, `std::stof` instead of boost::lexical_cast

Investigate any other boost libraries which trigger warnings.

#### Description:
A few of the boost libraries use `sprintf`; this triggers compiler spew, e.g.

```
stan/lib/stan_math/lib/boost_1.78.0/boost/lexical_cast/detail/converter_lexical_streams.hpp:310:21: warning: 'sprintf' is deprecated: This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of sprintf(3), it is highly recommended that you use snprintf(3) instead. [-Wdeprecated-declarations]
```

In the case of `boost::lexical_cast`, used to parse command line arguments, replacing it with `std::stoi` and `std::stof` etc would be faster.



#### Reproducible Steps:

compile CmdStan with clang v.14

```
clang --version
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
```

#### Current Output:
See above

#### Expected Output:
Code should compile without warnings triggered by libraries we use.




#### Current Version:
v2.32.2

--- a/src/cmdstan/arguments/singleton_argument.hpp
+++ b/src/cmdstan/arguments/singleton_argument.hpp
@@ -2,12 +2,38 @@
 #define CMDSTAN_ARGUMENTS_SINGLETON_ARGUMENT_HPP
 
 #include <cmdstan/arguments/valued_argument.hpp>
-#include <boost/lexical_cast.hpp>
 #include <iostream>
 #include <string>
 #include <vector>
 
 namespace cmdstan {
+
+namespace internal {
+void from_string(std::string &src, double &dest) { dest = std::stod(src); }
+void from_string(std::string &src, int &dest) { dest = std::stoi(src); }
+void from_string(std::string &src, long long int &dest) {
+  dest = std::stoll(src);
+}
+void from_string(std::string &src, unsigned int &dest) {
+  dest = std::stoul(src);
+}
+void from_string(std::string &src, bool &dest) {
+  if (src == "true" || src == "1") {
+    dest = true;
+  } else if (src == "false" || src == "0") {
+    dest = false;
+  } else {
+    throw std::invalid_argument(std::string("invalid boolean value ") + src);
+  }
+}
+void from_string(std::string &src, std::string &dest) { dest = src; }
+
+std::string to_string(std::string &src) { return src; }
+template <typename T>
+std::string to_string(T &src) {
+  return std::to_string(src);
+}
+}  // namespace internal
 
 template <typename T>
 struct type_name {
@@ -73,7 +99,8 @@ class singleton_argument : public valued_argument {
       args.pop_back();
 
       try {
-        T proposed_value = boost::lexical_cast<T>(value);
+        T proposed_value;
+        internal::from_string(value, proposed_value);
         if (set_value(proposed_value)) {
           return true;
         }
@@ -109,7 +136,7 @@ class singleton_argument : public valued_argument {
     return false;
   }
 
-  std::string print_value() { return boost::lexical_cast<std::string>(_value); }
+  std::string print_value() { return internal::to_string(_value); }
 
   std::string print_valid() { return " " + _validity; }
 

--- a/src/test/interface/arguments/singleton_argument_test.cpp
+++ b/src/test/interface/arguments/singleton_argument_test.cpp
@@ -1,6 +1,7 @@
 #include <cmdstan/arguments/singleton_argument.hpp>
 #include <stan/callbacks/stream_writer.hpp>
 #include <stan/callbacks/writer.hpp>
+#include <boost/lexical_cast.hpp>
 #include <gtest/gtest.h>
 
 template <typename T>

